File: Problem-3.js
//Time: O(log n), Space: O(1)

var findPeakElement = function(nums) {
    let l = 0, r = nums.length - 1;
       while (l < r) {
           let mid = l + Math.floor((r-l) / 2);
           if (nums[mid] > nums[mid + 1])
               r = mid;
           else
               l = mid + 1;
       }
       return l;
};

File: problem_3.cpp
// Time Complexity : O(logn)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        int l = 0;
        int h = nums.size() - 1;
        
        while (l <= h) {
            int m = l + (h - l) / 2;
            if (nums[m] > nums[m-1] && nums[m] > nums[m+1]) return m;
            else if (m == 0 || nums[m-1] < nums[m]) {
                //move right, possiblity of finding peak is more
                l = m + 1;
            } else {
                h = m - 1;
            }
        }
        
        return -1;
    }
};
